feat: support extract content api

xnoname79 opened the pull request:
install links command
add content service

xnoname79 committed:
feat: support extract content api
install links command
add content service

TaQuangTu commented on examples/client.go:
```diff
@@ -17,39 +18,22 @@ func main() {
 	defer conn.Close()
 
-	c := protogen.NewSearchServiceClient(conn)
+	c := contentsvc.NewContentServiceClient(conn)
```
Don't remove the client for the search function, create another file for content extraction function instead.

TaQuangTu commented on internal/server/content_server.go:
```diff
@@ -0,0 +1,103 @@
+		eg, egCtx = errgroup.WithContext(ctx)
+	)
+	for _, link := range request.Links {
+		link := link
```
Consider removing this copy cause we are using Go 1.24 (>=1.22), in which loop variables are copied for each iteration already, see [example](https://go.dev/blog/loopvar-preview)

TaQuangTu commented on internal/server/content_server.go:
```diff
@@ -0,0 +1,103 @@
+	}
+	var (
+		response = &contentsvc.ExtractContentFromLinksReponse{
+			Contents: make([]*contentsvc.ExtractedContent, 0),
```
Cause we know the size in advanced, pre-allocate the array to avoid array expanding performance issue:
Use `make([]*contentsvc.ExtractedContent, 0, len(request.Links))`

xnoname79 committed:
fix: cmd scrap html
change links -> curl
consider replacing curl by go-colly
fix Dockerfile

xnoname79 committed:
fix: support semaphore lock for concurrency
add semaphore lock for content extraction process
add mutex lock for each request resource

xnoname79 committed:
fix: semaphore lock
correct using semaphore lock
change mutex lock to sync.Mutex for single request

xnoname79 committed:
fix: content extraction
replace link command by colly

TaQuangTu committed:
feat: update content server

TaQuangTu committed:
docs: update readme and example for content extraction from links

xnoname79 commented on internal/server/content_server.go:
```diff
@@ -0,0 +1,140 @@
+		if err != nil {
+			return nil, err
+		}
+		index := index
```
we can remove this line

TaQuangTu replied:
index := index you mean?

xnoname79 commented on internal/server/content_server.go:
```diff
@@ -0,0 +1,140 @@
+			}
+
+			mutex.Lock()
+			defer mutex.Unlock()
```
As we've used index already, we can remove this local lock

TaQuangTu committed:
feat: remove mutex
